scarlet/fft.py

```diff
@@ -0,0 +1,297 @@
+import operator
+
+import autograd.numpy as np
+from scipy import fftpack
+
+
+def _centered(arr, newshape):
+    """Return the center newshape portion of the array.
+
+    This function is used by `fft_convolve` to remove
+    the zero padded region of the convolution.
+
+    Note: If the array shape is odd and the target is even,
+    the center of `arr` is shifted to the center-right
+    pixel position.
+    This is slightly different than the scipy implementation,
+    which uses the center-left pixel for the array center.
+    The reason for the difference is that we have
+    adopted the convention of `np.fft.fftshift` in order
+    to make sure that changing back and forth from
+    fft standard order (0 frequency and position is
+    in the bottom left) to 0 position in the center.
+    """
+    newshape = np.asarray(newshape)
+    currshape = np.array(arr.shape)
+
+    if not np.all(newshape <= currshape):
+        msg = "arr must be larger than newshape in both dimensions, received {0}, and {1}"
+        raise ValueError(msg.format(arr.shape, newshape))
+
+    startind = (currshape - newshape+1) // 2
+    endind = startind + newshape
+    myslice = [slice(startind[k], endind[k]) for k in range(len(endind))]
+
+    return arr[tuple(myslice)]
+
+
+def _pad(arr, newshape, axes=None):
+    """Pad an array to fit into newshape
+
+    Pad `arr` with zeros to fit into newshape,
+    which uses the `np.fft.fftshift` convention of moving
+    the center pixel of `arr` (if `arr.shape` is odd) to
+    the center-right pixel in an even shaped `newshape`.
+    """
+    if axes is None:
+        newshape = np.asarray(newshape)
+        currshape = np.array(arr.shape)
+        dS = newshape - currshape
+        startind = (dS+1) // 2
+        endind = dS - startind
+        pad_width = list(zip(startind, endind))
+    else:
+        # only pad the axes that will be transformed
+        pad_width = [(0, 0) for axis in arr.shape]
+        try:
+            len(axes)
+        except TypeError:
+            axes = [axes]
+        for a, axis in enumerate(axes):
+            dS = newshape[a] - arr.shape[axis]
+            startind = (dS+1) // 2
+            endind = dS - startind
+            pad_width[axis] = (startind, endind)
+    return np.pad(arr, pad_width, mode="constant")
+
+
+def _get_fft_shape(img1, img2, padding=3, axes=None):
```

Comment:
Way back when we first started to doing FFTs I did some testing using different padding and it seemed like adding the two shapes with an additional 3 pixels provided the best pixel area to suppression of artifacts ratio, so I've been using it as the default padding ever since. I don't remember the exact number, but I do recall it saved a few orders of magnitude in the residuals over no padding.